File: projects/22-smart-waste-management/fog/aggregation.js
"use strict";

// Pure window-aggregate math, run once per group at flush time over
// whatever readings the double buffer handed back -- never incrementally
// maintained during ingest. Aggregate fields are count/min/max/avg/latest;
// latest is last-in-order (the final element of the readings array), not
// the max-timestamp value.
function summarizeWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((acc, v) => acc + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { summarizeWindow };

File: projects/12-smart-building-energy/fog/aggregation.py
from statistics import fmean


def summarise_window(sensor_type, site_id, unit, readings, window_start, window_end):
    """Reduce one window's raw {"ts","value"} readings into a single summary record; latest is the last-arrived reading, not the max-timestamp one."""
    sample_values = [r["value"] for r in readings]
    return {
        "sensor_type": sensor_type,
        "site_id": site_id,
        "unit": unit,
        "window_start": window_start,
        "window_end": window_end,
        "count": len(sample_values),
        "min": min(sample_values),
        "max": max(sample_values),
        "avg": round(fmean(sample_values), 3),
        "latest": readings[-1]["value"],
    }
